Award sorting

karllindmark opened the pull request:
@peter-budo:

OK, this PR can be seen as the "guidance" for how the other filters will be implemented. It's possible that we'll have to create some sort of base fragment to manage how the sorting works, but I think that's a future thing to look at when we have the bigger picture (ie, more than just the awards grid).

Commit:
Add basic award sorting

Commit:
Now possible to filter awards on category [rough implementation]

Commit:
Extract common functionality into base class

Commit:
Add state keeping

Commit:
Extract menu strings into strings.xml

Commit:
Update changelog

Commit:
Rename method from filterAwards(String) to filterItems(String)

peter-budo commented on BF4Intel/src/main/java/com/ninetwozero/bf4intel/database/dao/awards/AwardSorter.java:
```diff
@@ -13,31 +14,46 @@
 
-    public static SortedAwardContainer sort(final Awards awards) {
-        return awards != null ? fetchSortedAwards(awards) : new SortedAwardContainer();
+    private static SortedAwardContainer sortItemsByProgress(final Awards awards) {
```
I think that `sortItemsByProgess` can be simple renamed to `sortByProgress` and same for `sortItemsByCategory` to `sortByCategory` I see no reason for `Item` in name as action by progress/category including parameters makes it clear what these methods do

peter-budo committed:
Merge branch 'dev' into award-sorting

peter-budo commented on BF4Intel/src/main/res/values/strings.xml:
```diff
@@ -213,7 +213,8 @@
   <string name="msg_user_said_hooah">You said hooah</string>
-  <string name="label_sort_categorized">By category</string>
+  <string name="label_sort_all">Show all</string>
+  <string name="label_sort_categorized">By category&#8230;</string>
```
Any specific reason for 3 dots at the end?

peter-budo commented:
What about menu layout like this???

```
<?xml version="1.0" encoding="utf-8"?>

<menu xmlns:android="http://schemas.android.com/apk/res/android">
  <item
    android:id="@+id/ab_action_sort"
    android:icon="@drawable/ic_action_sort_1"
    android:title="@string/menu_title_sort"
    android:showAsAction="always">
    <menu>
      <group
        android:id="@+id/menu_group_sort"
        android:checkableBehavior="single">
        <item
          android:id="@+id/ab_action_sort_all"
          android:title="@string/label_sort_all"
          android:showAsAction="never"
          android:icon="@null" />
        <item
          android:id="@+id/ab_action_sort_progress"
          android:title="@string/label_sort_progress"
          android:showAsAction="never"
          android:icon="@null" />
        <item
          android:id="@+id/ab_action_sort_categorized"
          android:title="@string/label_sort_categorized"
          android:showAsAction="never"
          android:checkable="false"
          android:icon="@null" />
        <item
          android:id="@+id/ab_action_sort_cat_kits"
          android:title="@string/kits"
          android:showAsAction="never"
          android:icon="@null" />
        <item
          android:id="@+id/ab_action_sort_cat_game_modes"
          android:title="@string/game_modes"
          android:showAsAction="never"
          android:icon="@null" />
        <item
          android:id="@+id/ab_action_sort_cat_weapons"
          android:title="@string/weapons"
          android:showAsAction="never"
          android:icon="@null" />
        <item
          android:id="@+id/ab_action_sort_cat_vehicles"
          android:title="@string/vehicles"
          android:showAsAction="never"
          android:icon="@null" />
        <item
          android:id="@+id/ab_action_sort_cat_team"
          android:title="@string/team"
          android:showAsAction="never"
          android:icon="@null" />
      </group>
    </menu>
  </item>
</menu>
```

peter-budo commented:
Also it would be good idea to set selected option to action bar sub title. Any thoughts on this?

peter-budo committed:
• Provided AbstractSorter that will be extended by all sorter classes
 • Added constructor for AwardSorter in favour of static method calls

peter-budo committed:
Merge branch 'dev' into award-sorting

peter-budo committed:
Merge branch 'award-sorting-suggestion' into award-sorting

peter-budo committed:
• Renamed By category to Select category
 • Moved By progress above Select category
 • Added setActionBarSubTitle to BaseFragment
 • Updating action bar subtitle on menus selections

peter-budo commented:
@karllindmark 
- merged travis update
- merged my suggestions
- added setActionBarSubTitle method in BaseFragment

TODO:
- add sub title save state as it is reseted on rotation

peter-budo commented:
@karllindmark 
I just realized that would be good if we had floating menu rather one attached to action bar because unlocks for weapons has 11 categories. Scrollable menu attached to action bar is bad

karllindmark commented:
@peter-budo: In other words, a `ListView` in a `DialogFragment`?

peter-budo commented:
Yes that is what I was thinking. However just now been looking on Share in photo album.
1. User click on share icon in AB
2. Small menu is showed, with `See all`
3. User click on `See all`
4. Menu get updated with full scrollable list, see all is removed

We could apply it to our case
1. Click filter icon in AB
2. Show `Show all, By progress, By category`
3. User clicks `By category`
4. Menu get updated and lists `Show all, By progress, Kit, Weapons, Vehicles, etc` , but by category is removed

Commit:
Update menu to remove indicator

karllindmark commented:
@peter-budo: Current TODO as I see it:
1. Fix the subtitle glitches (state handling, initial display)
2. Suggestion: "Show all" --> "Default", "Select category" --> "By category" (to match the progress)
3. Question: How do we want to present the categories? It's a bit of a hassle to implement that "menu expansion" as far as I can tell (see the link on Hangouts).

I might've missed something though, but I think these three things may be left. :-)

peter-budo commented:
@karllindmark 
1. We could fix that with `Fragment.setRetainInstance(boolean retain)` however for some reason it is already set to false in BaseFragment [line45](https://github.com/ninetwozero/com.ninetwozero.bf4intel/blob/dev/BF4Intel/src/main/java/com/ninetwozero/bf4intel/base/ui/BaseFragment.java#L45) I presume you had reason for doing this
2. I agree with "Default", however in place of Select category/By category I would suggest "Show categories" to mimic "Show all" from Share context menu
3. I just spoken with guys here and they suggest simple logical approach, on "Show categories"  click hide this option and dynamically add all categories (kit, weapon, etc). This could possibly be bonus as we would have "default" menu with default+progress+categories, to which we would dynamically add another section specific menu

karllindmark commented:
@peter-budo: 
1. Actually, it's set to true (!false) if it doesn't find that "disable" flag that's used for nested fragments
2. I currently agree, until further notice. :space_invader: 
3. The main problem I noticed is that when you select a regular menu option (ie, "Select category" at the moment) it will automatically dismiss the dropdown menu, so the user will have to re-open it to view the new structure (since we also have to invalidate the menu). Could you ask your colleagues how they'd solve that issue? It's merely not just an issue of "hiding and showing", it's more of how to keep the dropdown visible after selecting "Select category".

I do like the idea of having a generic menu XML file, with the addition of dynamic population. :+1:

peter-budo commented:
> Suggestion: "Show all" --> "Default"

Actually thinking of it again I think default will be sort of undescriptive. What about "Show all - default" or "All by default"?

Also I tried `Fragment.setRetainInstance(boolean retain)` as true but subtitle still got reset. Will need to research more...

peter-budo committed:
• Simplified context menus logic

peter-budo commented:
I guess this one can be closed given that work continues on `sort-menu-provider` PR
